File: sales-service/src/main/java/com/aalvarenga/sales/purchase/CardExpiration.java
package com.aalvarenga.sales.purchase;

import java.time.YearMonth;

/**
 * Validade de cartão no formato MM/YY (anos 2000-2099), mesma regra do billing:
 * mês entre 01 e 12 e não vencido (um cartão "12/26" vale até o fim de dez/2026).
 */
public final class CardExpiration {

    private CardExpiration() {
    }

    public static boolean isValid(String expiration, YearMonth currentMonth) {
        if (expiration == null || !expiration.matches("(0[1-9]|1[0-2])/\\d{2}")) {
            return false;
        }
        int month = Integer.parseInt(expiration.substring(0, 2));
        int year = 2000 + Integer.parseInt(expiration.substring(3));
        return !YearMonth.of(year, month).isBefore(currentMonth);
    }
}

File: sales-web/src/lib/cardExpiration.ts
/**
 * Validade MM/AA (anos 2000-2099), mesma regra do billing e do backend:
 * mês 01-12 e não vencido (um cartão 09/26 vale durante todo setembro/2026).
 */
export function isValidExpiration(expiration: string, today: Date = new Date()): boolean {
  const match = /^(0[1-9]|1[0-2])\/(\d{2})$/.exec(expiration);
  if (!match) {
    return false;
  }
  const month = Number(match[1]);
  const year = 2000 + Number(match[2]);
  const currentYear = today.getFullYear();
  const currentMonth = today.getMonth() + 1;
  return year > currentYear || (year === currentYear && month >= currentMonth);
}

/** Máscara enquanto digita: "0331" -> "03/31". */
export function maskExpiration(value: string): string {
  const digits = value.replace(/\D/g, '').slice(0, 4);
  return digits.length > 2 ? `${digits.slice(0, 2)}/${digits.slice(2)}` : digits;
}
